### ISSUE TITLE
Axiom on computer division instead of definition makes VC unprovable 

### ISSUE THREAD

**Issue by yannickmoy** [cvc5/cvc5#9427]
**Describe the bug**

I'm investigating if we could better define computer division for cvc5. Currently, we're using the axioms defined in Why3 for computer division, which ensure that the rest of the division is of the sign of the dividend.

When using a definition, cvc5 can prove a property about computer division and abs:

```smt
(set-logic AUFBVFPDTNIRA)

(declare-const x Int)
(declare-const y Int)

(define-fun cdiv ((x Int) (y Int)) Int
  (ite (or (< 0 x) (= (mod x y) 0))
    (div x y)
    (ite (< 0 y)
      (+ (div x y) 1)
      (- (div x y) 1))))

(assert (not (= y 0)))
(assert (not (= (abs (cdiv x y)) (cdiv (abs x) (abs y)))))

(check-sat)
```

but if I replace this definition with a defining axiom (and pattern), cvc5 does not return:

```smt
(set-logic AUFBVFPDTNIRA)

(declare-const x Int)
(declare-const y Int)

(declare-fun cdiv (Int Int) Int)

(assert
  (forall ((x Int) (y Int))
    (! (=> (not (= y 0)) (= (cdiv x y)
  (ite (or (< 0 x) (= (mod x y) 0))
    (div x y)
    (ite (< 0 y)
      (+ (div x y) 1)
      (- (div x y) 1))))) :pattern ((cdiv x y)))))

(assert (not (= y 0)))
(assert (not (= (abs (cdiv x y)) (cdiv (abs x) (abs y)))))

(check-sat)
```

Could you explain why?
That may inform other adaptations of VCs for cvc5. 
Thanks

**Command line arguments**: none 
**cvc5 version/commit**: 1.0.3
**Operating system**: Linux

**Comment by martin-cs** [comment]
On Wed, 2023-01-18 at 06:59 -0800, yannickmoy wrote:
> Could you explain why?
> That may inform other adaptations of VCs for cvc5.

I may be a little out-of-date but last I recall, (non-recursive)
definition functions were just "inlined" / expanded while the
axiomatisation would be passed to the quantifier theory solver and so
have to go through all of the checks, hoops and complications of that.

In terms of general sanity, I would strongly suggest the definitional
form.

**Comment by yannickmoy** [comment]
I agree with you in general. Regarding non-linear operations, the results are better with axiomatized definitions on many VCs. This is why Why3 still chooses to axiomatize computer division for cvc5 currently (with axioms that don't give a definition like the one above). 

As an example where the above axiom is better than the definition, see this VC:

```smt
(set-logic AUFBVFPDTNIRA)

(declare-const x Int)
(declare-const y Int)

(define-fun cdiv ((x Int) (y Int)) Int
  (ite (or (< 0 x) (= (mod x y) 0))
    (div x y)
    (ite (< 0 y)
      (+ (div x y) 1)
      (- (div x y) 1))))

(assert (not (= y 0)))
(assert (not (= (cdiv (- x) y) (- (cdiv x y)))))

(check-sat)
```

It is not proved by cvc5.
But if I replace the definition of `cdiv` with the equivalent axiom, it is proved by cvc5.

**Comment by yannickmoy** [comment]
Another interesting observation is that the following alternative definition of `cdiv` is much worse for cvc5:

```smt
(define-fun cdiv ((x Int) (y Int)) Int
  (ite (<= 0 x)
    (div x y)
    (- (div (- x) y))))
```

cvc5 still proves VCs with that definition, but it takes much longer.
which makes me think that I'd be interested in your advice on an optimal definition of computer division.

**Comment by yannickmoy** [comment]
By trial and error, I got this definition which allows to prove quickly both VCs presented before:

```smt
(define-fun cdiv ((x Int) (y Int)) Int
  (ite (or (< 0 x) (= (mod x y) 0))
    (div x y)
    (ite (< 0 y)
      (- (div (- x) y))
      (- (div (- x) y)))))
```

Obviously not the simplest definition, but I'm willing to use the definition that makes cvc5 happiest!

**Comment by ajreynol** [comment]
For variants of these benchmarks where cvc5 does not solve, many cases it is due to infinite branch-and-bound lemmas being emitted from the quantifier-free linear integer solver.

In several cases, adding option `--decision=internal` appears to make cvc5 solve the benchmark quickly, including the quantified version of the first 2 definitions of cdiv.  However, it goes solved -> unsolved on the quantifier-free version of the 4th variant.

The solution here is likely to make us less sensitive to the branch and bound heuristics in the linear integer solver.  We should revisit our interleaving of non-linear + quantified reasoning with linear reasoning.  I will keep you posted if I'm able to improve things on benchmarks from this issue.

**Comment by yannickmoy** [comment]
Thanks @ajreynol.

FYI, we adopted the following definition for computer division, and similar ones for mod/rem (that are different operations in SPARK Ada), and removed all axioms, as the results were much better with cvc5 for both proofs and models:

```smt
(define-fun cdiv ((x Int) (y Int)) Int
  (ite (<= 0 x)
    (div x y)
    (- (div (- x) y))))
```

I will try out `--decision=internal` on out testsuite, but initial experiments on arithmetic tests showed regressions.

**Comment by yannickmoy** [comment]
FYI, I finally got to run the SPARK testsuite with `--decision=internal`, and the results are mixed, with more regressions than improvements.
